Rework configuration parsing (again)

DaanDeMeyer opened the pull request:
As explained in https://github.com/systemd/mkosi/issues/2846, there are multiple issues with the current
implementation of mkosi.images. Let's take what we learned from the
default initrd and the default tools tree and apply it mkosi.images.

Specifically, all issues arise from the fact that we apply every option
from the global configuration (including CLI arguments) to the images
from mkosi.images/. To avoid the issues that arise from this (e.g --package
abc installing abc in all images), we made configuration values override
CLI arguments again so that we could override faulty CLI arguments again
in subimages so that they would only apply to the main image (e.g. set
Format= explicitly for each subimage so that --format on the command line
only applies to the main image).

Because we still wanted to allow configurable settings that can be modified
via the command line, we introduced the default specifier '@' which can be
prefixed to a setting to set a default value instead of overriding the value.
The '@' specifier is generally used in the global image independent
configuration to specify default values that can be overridden from the command
line. This specifier has led to a lot of confusion, along with the behavior that
the CLI does not override the configuration.

From the default tools tree and default initrd, we learned that what works
very well is to only have specific settings from the main image configuration
apply to the default tools tree and default initrd. For example, the distribution,
release, mirror and architecture should be the same for the main image and the
initrd, but the packages from the main image should not all be installed in the
initrd.

We can apply this idea to the images from mkosi.images/ as well, if we
introduce the assumption that all images defined in mkosi.images are
subimages intended to be included in some way or form in the main image.
This assumption allows us to divide all settings into either image specific
settings or "universal" settings that should apply to the main image and all
its subimages. The universal settings are passed on to each subimage. The image
specific settings are not.

This idea also allows us to define the "main" image outside of mkosi.images
again. Since only "universal" settings are passed on, we can safely define
an output format and such again in the global configuration, as we know this
won't be passed on to subimages.

It also allows us to make CLI arguments override configuration again. Since
there is no need anymore for subimages to override the CLI configuration as
inappropriate CLI configuration such as extra packages will only apply to the
main image and not any subimages from mkosi.images/. Because CLI configuration
overrides file configuration again, we also don't need the '@' specifier anymore,
as default values can simply be set without '@', since the CLI will override
the configuration file values by default.

We also lose the need for --append, because the sole use for --append was again
to override file based configuration.

This commit implements all of what's mentioned above, specifically:

- CLI configuration now always trumps file based configuration.
- The '@' specifier is dropped automatically during parsing
- The main image is now always added from global configuration, even
if there are images in mkosi.images/. The main image is always built
last, and cannot be used as a dependency in the Dependencies= setting
for images defined in mkosi.images/.
- The Dependencies= setting for the main image now is used to specify
which subimages from mkosi.images/ to build. By default all subimages
are built.
- A universal tag is introduced for settings and appropriate settings
are marked as universal. Universal settings are passed on from the
main image configuration to subimage configuration.
- The Images= setting is removed, as it's role is replaced by
Dependencies=.
- The old name mkosi.presets and the Preset section name are removed
as they have been deprecated for a long time now.
- The config parsing tests are extended to cover more cases.
- All builtin configuration is adapted to stop using the '@' specifier.
- The documentation is updated in accordance with the changes.

DaanDeMeyer commented:
This breaks backwards compat in the following ways:

- `@` is removed, fix is to just remove the `@`. We can do this in mkosi itself and log a warning
- `Images` is removed, which is only applicable when building multiple images with `mkosi.images/`, which is rare
- Main image is now always included even when `mkosi.images` is used. Fix is to move the main image configuration from `mkosi.images` to the global configuration.
- `--append` is removed, which is only used by the systemd integration tests I think
- `mkosi.images` can only be used to build subimages now, with more restrictions put on them (same distribution, release, mirror, ...). Not all CLI settings apply to `mkosi.images/` anymore (which is a good thing).

Anyone building single images without using `mkosi.images/` will not be affected if we automatically remove `@` from settings within mkosi. This makes me think that breaking compat here is worth it given the substantial simplification of config parsing we get from it.

NekkoDroid commented:
Just went to test it out (adjusting my config for this) and noticed that `%i` was being ignored. `ImageId` and `ImageVersion` probably should also have the universal tag applied.

DaanDeMeyer commented:
> Just went to test it out (adjusting my config for this) and noticed that `%i` was being ignored in subimages. `ImageId` and `ImageVersion` probably should also have the universal tag applied.

Added both as universal settings.

> 1. Is there any reason to keep the symlinks for subimages around now that they aren't used for qemu/shell/...? They can't really be used as base trees for anything since the mounting depends on the extension.

I've fixed that now by making sure we resolve symlinks in `mount_base_trees()`

> 2. Is there some way to apply a global `RemoveFiles=` list to the image & subimages? I was using it globally to clean up directories & files that are shipped as part of `Packages=` and are needed for builds to work, but aren't needed any output image.

Have a shared snippet with the `RemoveFiles=` and `Include=` it in every subimage.

DaanDeMeyer commented:
Also changed the behavior to automatically drop the '@' specifier in mkosi itself to reduce the breakage caused by this PR

behrmann commented on mkosi/config.py:
```diff
@@ -3409,25 +3431,51 @@ def __call__(
+            # To determine default values, we need the final values of various settings in
+            # a namespace object, but we don't want to copy the final values into the config
+            # namespace object just yet so we create a new namespace object instead.
+            factoryns = argparse.Namespace(**{
+                d: finalize_value(SETTINGS_LOOKUP_BY_DEST[d]) for d in setting.default_factory_depends
+            })
```
I think this way it's easier to not overlook that unpacking is happening.
```suggestion
            factoryns = argparse.Namespace(
                **{d: finalize_value(SETTINGS_LOOKUP_BY_DEST[d]) for d in setting.default_factory_depends}
            )
```

DaanDeMeyer commented on mkosi/config.py:
```diff
@@ -730,6 +727,24 @@ def config_default_proxy_url(namespace: argparse.Namespace) -> Optional[str]:
+
+    dependencies = []
+
+    for p in sorted(d.iterdir()):
```
Yes but then we do an early return so it's OK?

behrmann replied:
Right, true. I think what I trips me here is, the thing is static, so instead of the inline definition I think something like this would be more readable:
```python
    imagedir = Path("mkosi.images")
    if namespace.directory is None or not imagedir.exists():
        return []

    if namespace.image:
        return []

    dependencies = []

    for p in sorted(imagedir.iterdir()):
```

DaanDeMeyer commented on mkosi/config.py:
```diff
@@ -3293,12 +3315,16 @@ def resolve_deps(images: Sequence[argparse.Namespace], include: Sequence[str]) -
-    immutable_settings: set[str] = set()
-    append = False
+
+    class State:
```
Renamed to `ParseContext`

We can't make it a dataclass that we instantiate anew every time because the inner functions would not get the new object but still operate on the old one

DaanDeMeyer commented:
Added one more fix to make sure that `mkosi.local.conf` still overrides configuration that is parsed later.

behrmann commented on mkosi/config.py:
```diff
@@ -3610,108 +3674,90 @@ def finalize_defaults() -> None:
+    images = []
 
-    append = True
+    if args.directory is not None and (d := Path("mkosi.images")).exists():
```
That too would be more readable as
```suggestion
    if args.directory is not None and Path("mkosi.images").exists():
```
and the one usage of `d` below changed as in the other place.

septatrix commented:
> * Main image is now always included even when `mkosi.images` is used. Fix is to move the main image configuration from `mkosi.images` to the global configuration.
> * `mkosi.images` can only be used to build subimages now, with more restrictions put on them (same distribution, release, mirror, ...). Not all CLI settings apply to `mkosi.images/` anymore (which is a good thing).

I have two images which are strongly related but neither is a subimage or dependency of the other. What is the best way to have such a configuration with the new config mechanism?

More specifically I have a system image and a separate installer. If I move the former to the main image I will not be able to build the latter separate if I understand correctly (or at least I have no clue what to pass on the command line to only build the installer).

DaanDeMeyer commented:
> > * Main image is now always included even when `mkosi.images` is used. Fix is to move the main image configuration from `mkosi.images` to the global configuration.
> > * `mkosi.images` can only be used to build subimages now, with more restrictions put on them (same distribution, release, mirror, ...). Not all CLI settings apply to `mkosi.images/` anymore (which is a good thing).
> 
> I have two images which are strongly related but neither is a subimage or dependency of the other. What is the best way to have such a configuration with the new config mechanism?
> 
> More specifically I have a system image and a separate installer. If I move the former to the main image I will not be able to build the latter separate if I understand correctly (or at least I have no clue what to pass on the command line to only build the installer).

@septatrix My suggestion here is to have two different configuration which you build separate using `mkosi --directory <...>`. Shared configuration between the two images can be handled with `Include=`.

septatrix commented:
> @septatrix My suggestion here is to have two different configuration which you build separate using `mkosi --directory <...>`. Shared configuration between the two images can be handled with `Include=`.

I came up with a different solution in the meantime. I set `Format=none` for the main image (which should completely ignore it?) and use a profile to swap out the `Dependencies=` as well as a `[Match]` in the system image which checks for the profile and adjusts `Format=`, `SplitArtifacts=` accordingly. However, I am not sure if this is intended or bound to be broken at some later point.

The directory approach seems a bit annoying because in the installer I need to reference the split artifacts from the system image

septatrix commented:
Also: Will `mkosi qemu/boot` now always start the main image? Using `mkosi.images` but not being able to start nothing but the main image seems a bit of a misnomer if that limitation really exists...

DaanDeMeyer commented:
@septatrix Yes we'll always start the main image now. We had to introduce limitations to the feature to make it usable. We're not aiming for mkosi to be a general purpose build system. If mkosi.images doesn't fit your needs, you'll need to add something on top to orchestrate multiple image builds.

septatrix commented:
Okay, in that case I will need to resort to `-C`, understood. It just feels like `mkosi.images` is now a misnomer. Similarly `mkosi summary` still talk about "default" images but if there is no way to start different images that might be confusing. The way this directory is now handled is more akin to `mkosi.components` or something like it because the sub"images" are intended to be used standalone but instead are more like glorified build scripts

septatrix commented:
It also seems like I need a separate `mkosi.version`, `mkosi.crt`, and `mkosi.key` files for each of them then... :/

DaanDeMeyer commented:
> It also seems like I need a separate `mkosi.version`, `mkosi.crt`, and `mkosi.key` files for each of them then... :/

`Include=` is your friend, you can have a directory containing all of those and simply `Include=` it wherever needed.

septatrix commented:
> > It also seems like I need a separate `mkosi.version`, `mkosi.crt`, and `mkosi.key` files for each of them then... :/
> 
> `Include=` is your friend, you can have a directory containing all of those and simply `Include=` it wherever needed.

This works only to a certain degree. `mkosi bump` does not work and `mkosi genkey` still places it in the individual directories (even when settings `SshCertificate=`). It is better than nothing but its fair to say that my use case and current workflow has definitely been hit by this change (though I can understand why it was done)

DaanDeMeyer commented:
> > > It also seems like I need a separate `mkosi.version`, `mkosi.crt`, and `mkosi.key` files for each of them then... :/
> > 
> > 
> > `Include=` is your friend, you can have a directory containing all of those and simply `Include=` it wherever needed.
> 
> This works only to a certain degree. `mkosi bump` does not work and `mkosi genkey` still places it in the individual directories (even when settings `SshCertificate=`). It is better than nothing but its fair to say that my use case and current workflow has definitely been hit by this change (though I can understand why it was done)

Yeah I'm sorry about that, but we had to consider all the users not using mkosi.images as well
